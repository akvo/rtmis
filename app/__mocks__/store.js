export const UIState = {
  update: jest.fn(),
  useState: jest.fn(),
};

export const UserState = {
  update: jest.fn(),
  useState: jest.fn(),
};

export const AuthState = {
  update: jest.fn(),
  useState: jest.fn(),
};

export const FormState = {
  update: jest.fn(),
  useState: jest.fn(),
};

export const BuildParamsState = {
  update: jest.fn(),
  useState: jest.fn(),
};
