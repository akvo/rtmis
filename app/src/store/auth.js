import { Store } from 'pullstate';
/**
 * Server config
 */
const AuthState = new Store({
  authenticationCode: '',
  useAuthenticationCode: false, // using code for authentication
  username: '',
  password: '',
  token: null,
});

export default AuthState;
