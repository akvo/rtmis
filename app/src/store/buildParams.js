import { Store } from 'pullstate';
import defaultBuildParams from '../build';

const BuildParamsState = new Store({
  authenticationType: defaultBuildParams?.authenticationType || [
    'code_assignment',
    'username',
    'password',
  ],
  serverURL: defaultBuildParams?.serverURL,
  debugMode: defaultBuildParams?.debugMode || false,
  dataSyncInterval: defaultBuildParams?.dataSyncInterval || 300,
  errorHandling: defaultBuildParams?.errorHandling || true,
  loggingLevel: defaultBuildParams?.loggingLevel || 'verbose',
  appVersion: defaultBuildParams?.appVersion || '1.0.0',
  gpsThreshold: defaultBuildParams?.gpsThreshold || 20,
});

export default BuildParamsState;
