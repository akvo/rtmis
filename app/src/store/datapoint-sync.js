import { Store } from 'pullstate';

const DatapointSyncState = new Store({
  inProgress: false,
  progress: 0,
  added: false,
  completed: false,
});

export default DatapointSyncState;
