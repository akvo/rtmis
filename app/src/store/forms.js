import { Store } from 'pullstate';

const FormState = new Store({
  form: {},
  currentValues: {}, // answers
  questionGroupListCurrentValues: {}, // answers for question group list component
  visitedQuestionGroup: [], // to store visited question group id
  surveyDuration: 0,
  surveyStart: null,
  cascades: {},
  lang: 'en',
});

export default FormState;
