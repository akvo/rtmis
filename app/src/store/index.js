export { default as AuthState } from './auth';
export { default as BuildParamsState } from './buildParams';
export { default as FormState } from './forms';
export { default as UIState } from './ui';
export { default as UserState } from './users';
