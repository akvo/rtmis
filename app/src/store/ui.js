import { Store } from 'pullstate';

const UIState = new Store({
  isDarkMode: false, // if isDarkMode = false then Theme= light
  lang: 'en',
  fontSize: 16,
  currentPage: 'GetStarted',
  online: false,
  networkType: null,
  isManualSynced: false,
});

export default UIState;
