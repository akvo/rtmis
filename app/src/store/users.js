import { Store } from 'pullstate';

const UserState = new Store({
  id: null,
  name: '',
  password: '',
  token: null,
  syncWifiOnly: false,
  syncInterval: 300,
  forms: [],
});

export default UserState;
