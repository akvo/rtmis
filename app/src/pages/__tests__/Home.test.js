import React from 'react';
import { render, waitFor, fireEvent, act } from '@testing-library/react-native';
import HomePage from '../Home';
import crudForms from '../../database/crud/crud-forms';
import FormState from '../../store/forms';
import { UserState, UIState } from '../../store';

const mockDateNow = new Date().toISOString();
const mockForms = [
  {
    id: 1,
    formId: 9001,
    version: '1.0.0',
    latest: 1,
    name: 'Form 1',
    json: JSON.stringify({ id: 9001, name: 'Form 1', question_group: [] }),
    createdAt: mockDateNow,
    submitted: 2,
    draft: 0,
    synced: 2,
  },
  {
    id: 2,
    formId: 9002,
    version: '1.0.1',
    latest: 1,
    name: 'Form 2',
    json: JSON.stringify({ id: 9002, name: 'Form 2', question_group: [] }),
    createdAt: mockDateNow,
    submitted: 1,
    draft: 3,
    synced: 0,
  },
  {
    id: 3,
    formId: 9002,
    version: '1.0.0',
    latest: 0,
    name: 'Form 2',
    json: JSON.stringify({ id: 9002, name: 'Form 2', question_group: [] }),
    createdAt: mockDateNow,
    submitted: 2,
    draft: 1,
    synced: 1,
  },
];

jest.mock('../../database/crud/crud-forms');
jest.mock('../../store/forms');
const mockNavigation = {
  navigate: jest.fn(),
};

describe('Homepage', () => {
  beforeAll(() => {
    UserState.update((s) => {
      s.id = 1;
    });
    const mockLatestFormVersion = mockForms.filter((form) => form.latest);
    crudForms.selectLatestFormVersion.mockImplementation(() =>
      Promise.resolve(mockLatestFormVersion),
    );
  });

  test('renders correctly', async () => {
    const tree = render(<HomePage navigation={mockNavigation} />);

    await waitFor(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  it('should render page title, search field and back button', async () => {
    const wrapper = render(<HomePage navigation={mockNavigation} />);

    await waitFor(() => {
      const titleElement = wrapper.getByText('Form Lists');
      expect(titleElement).toBeDefined();

      const searchField = wrapper.getByTestId('search-bar');
      expect(searchField).toBeDefined();

      const backButton = wrapper.getByTestId('button-users');
      expect(backButton).toBeDefined();
    });
  });

  it('should load last form version data from DB with form stats', async () => {
    const wrapper = render(<HomePage navigation={mockNavigation} />);

    await waitFor(() => {
      expect(crudForms.selectLatestFormVersion).toHaveBeenCalledTimes(1);
    });

    const listForm1 = wrapper.queryByTestId('card-touchable-0');
    expect(listForm1).toBeTruthy();
    expect(listForm1.props.children[0].props.title).toEqual('Form 1');
    expect(listForm1.props.children[0].props.subTitles).toEqual([
      'Version: 1.0.0',
      'Submitted: 2',
      'Draft: 0',
      'Synced: 2',
    ]);

    const listForm2 = wrapper.queryByTestId('card-touchable-1');
    expect(listForm2).toBeTruthy();
    expect(listForm2.props.children[0].props.title).toEqual('Form 2');
    expect(listForm2.props.children[0].props.subTitles).toEqual([
      'Version: 1.0.1',
      'Submitted: 1',
      'Draft: 3',
      'Synced: 0',
    ]);

    const listForm3 = wrapper.queryByTestId('card-touchable-2');
    expect(listForm3).toBeFalsy();
  });

  it('should filter forms by search value', async () => {
    const wrapper = render(<HomePage navigation={mockNavigation} />);

    await waitFor(() => {
      expect(crudForms.selectLatestFormVersion).toHaveBeenCalledTimes(1);
    });

    const searchField = wrapper.getByTestId('search-bar');
    expect(searchField).toBeDefined();
    fireEvent.changeText(searchField, 'Form 1');

    const listForm1 = wrapper.queryByTestId('card-touchable-0');
    expect(listForm1).toBeTruthy();

    const listForm2 = wrapper.queryByTestId('card-touchable-1');
    expect(listForm2).toBeFalsy();

    const listForm3 = wrapper.queryByTestId('card-touchable-2');
    expect(listForm3).toBeFalsy();
  });

  it('should navigate to ManageForm page when form list pressed', async () => {
    const mockFindData = mockForms.find((form) => form.id === 1);
    FormState.update.mockImplementation(() => mockFindData);

    const wrapper = render(<HomePage navigation={mockNavigation} />);

    await waitFor(() => {
      expect(crudForms.selectLatestFormVersion).toHaveBeenCalledTimes(1);
    });

    const listForm1 = wrapper.queryByTestId('card-touchable-0');
    expect(listForm1).toBeTruthy();
    fireEvent.press(listForm1);

    await waitFor(() => {
      expect(mockNavigation.navigate).toHaveBeenCalledWith('ManageForm', { id: 1, name: 'Form 1' });
    });
  });

  it('should navigate to Users page when back button clicked', async () => {
    const wrapper = render(<HomePage navigation={mockNavigation} />);

    await waitFor(() => {
      expect(crudForms.selectLatestFormVersion).toHaveBeenCalledTimes(1);
    });

    const backButton = wrapper.getByTestId('button-users');
    expect(backButton).toBeDefined();
    fireEvent.press(backButton);

    await waitFor(() => {
      expect(mockNavigation.navigate).toHaveBeenCalledWith('Users');
    });
  });

  it('should reset form and data when app language changed', async () => {
    const { getByText, debug, queryByTestId } = render(<HomePage navigation={mockNavigation} />);

    act(() => {
      UIState.update((s) => {
        s.lang = 'fr';
      });
      FormState.update((s) => {
        s.form = {};
      });
    });

    await waitFor(() => {
      const listForm1 = queryByTestId('card-touchable-0');
      expect(listForm1).toBeTruthy();

      expect(listForm1.props.children[0].props.subTitles).toEqual([
        'Version: 1.0.0',
        'Soumis: 2',
        'Brouillon: 0',
        'Synchronis√©: 2',
      ]);
    });
  });
});
