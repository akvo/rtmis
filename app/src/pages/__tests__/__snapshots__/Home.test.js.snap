// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9fafb",
      "flex": 1,
      "flexDirection": "column",
      "flexWrap": "wrap",
    }
  }
  testID="stack-container"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f3f4f6",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "minheight": 78,
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="base-layout-page-title"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="button-users"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 2,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontSize": 18,
              "paddingTop": 4,
              "textAlign": "center",
            }
          }
          testID="page-title"
        >
          Form Lists
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="more-options-button"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 2,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
  <View
    style={
      {
        "backgroundColor": "#393e42",
        "borderBottomColor": "#000",
        "borderBottomWidth": 1,
        "borderTopColor": "#000",
        "borderTopWidth": 1,
        "padding": 8,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#303337",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 3,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              testID="RNE__ICON__CONTAINER_ACTION"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
        <TextInput
          editable={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Search form"
          placeholderTextColor="#86939e"
          platform="default"
          style={
            {
              "color": "#86939e",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
              "width": "100%",
            }
          }
          testID="search-bar"
          underlineColorAndroid="transparent"
          value={null}
        />
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "#f9fafb",
        "flex": 1,
        "flexDirection": "row",
        "flexWrap": "wrap",
      }
    }
    testID="stack-container"
  />
</View>
`;
